nbgitpuller/handlers.py

```diff
@@ -38,6 +40,38 @@ def emit(self, data):
         self.write('data: {}\n\n'.format(serialized_data))
         yield self.flush()
 
+    def setup_plugins(self, provider):
+        pm = pluggy.PluginManager("nbgitpuller")
+        pm.add_hookspecs(hookspecs)
+        pm.load_setuptools_entrypoints("nbgitpuller", name=provider)
+        return pm
+
+    @gen.coroutine
+    def progress_loop(self, queue):
+        while True:
+            try:
+                progress = queue.get_nowait()
+            except Empty:
+                yield gen.sleep(0.1)
```

Comment:
I did the gen.sleep(5) in order to allow the user to see the final messages in the progress window before being re-directed to the hub / notebook. I am not wedded to the five seconds but it seemed useful for the user to see the progress messages complete successfully or not. We can take the wait back down to a very small number of milliseconds if this does not seem useful.

Reply:
Oh, so the time when gen.sleep(5) is used, it is because its a one-off sleep until something is assumed to have resolved? I assumed it was until another attempt to see if something had changed.

Reply:
yes it is the moment just before the application redirects to the hub.